commander3/src/comm_tod_WMAP_mod.f90

```diff
@@ -19,469 +19,1237 @@
-     call wall_time(t1)
-     call self%output_scan_list(slist)
-     !TODO: all this timing stuff is a disaster
-     call wall_time(t2); t_tot(20) = t_tot(20) + t2-t1
-     call update_status(status, "scanlist2")
-   end if
-
-   ! Solve combined map, summed over all pixels
-   !TODO: this probably also needs changing, also why is this so long and not a
-   !functions
-   call wall_time(t1)
-   call update_status(status, "shared1")
-   if (sA_map%init) then
-     do i = 0, self%numprocs_shared-1
-      !call update_status(status, "tod_share_loop")
-      start_chunk = mod(self%myid_shared+i,self%numprocs_shared)*chunk_size
-      end_chunk   = min(start_chunk+chunk_size-1,npix-1)
-      do while (start_chunk < npix)
-         if (self%pix2ind(start_chunk) /= -1) exit
-         start_chunk = start_chunk+1
+      call update_status(status, "Running allreduce on M_diag")
+      call mpi_allreduce(mpi_in_place, M_diag, size(M_diag), &
+           & MPI_DOUBLE_PRECISION, MPI_SUM, self%info%comm, ierr)
+      call update_status(status, "Running allreduce on b")
+      call mpi_allreduce(mpi_in_place, b_map, size(b_map), &
+           & MPI_DOUBLE_PRECISION, MPI_SUM, self%info%comm, ierr)
+
+
+      np0 = self%info%np
+      allocate (cg_tot(0:np0 - 1, nmaps))
+
+      ! write out M_diag, b_map to fits.
+      cg_tot = b_map(self%info%pix, 1:nmaps, 1)
+      call write_fits_file_iqu(trim(prefix)//'b'//trim(postfix), cg_tot, outmaps)
+      !cg_tot = M_diag(self%info%pix, 1:nmaps)
+      !call write_fits_file_iqu(trim(prefix)//'M'//trim(postfix), cg_tot, outmaps)
+      !cg_tot = b_map(self%info%pix, 1:nmaps, 5)
+      !call write_fits_file_iqu(trim(prefix)//'b_sl'//trim(postfix), cg_tot, outmaps)
+
+
+      where (M_diag == 0d0)
+         M_diag = 1d0
+      end where
+
+      ! Conjugate Gradient solution to (P^T Ninv P) m = P^T Ninv d, or Ax = b
+      call update_status(status, "Allocating cg arrays")
+      allocate (cg_sol(0:npix-1, nmaps, nout))
+      if (self%myid == 0) then 
+         allocate (r     (0:npix-1, nmaps))
+         allocate (rhat  (0:npix-1, nmaps))
+         allocate (r0    (0:npix-1, nmaps))
+         allocate (q     (0:npix-1, nmaps))
+         allocate (p     (0:npix-1, nmaps))
+         allocate (s     (0:npix-1, nmaps))
+         allocate (shat  (0:npix-1, nmaps))
+         allocate (m_buf (0:npix-1, nmaps))
+         allocate (phat  (0:npix-1, nmaps))
+         allocate (v     (0:npix-1, nmaps))
+         allocate (determ(0:npix-1))
+         determ = M_diag(:,2)*M_diag(:,3) - M_diag(:,4)**2
+
+         cg_sol = 0.0d0
+         !cg_sol(:,:,1) = map_sky(:,:,0,1)
+         ! WMAP's metric was |Ax-b|/|b| < 10^-8, which essentially is 
+         epsil(1)   = 1d-10
+         epsil(2:6) = 1.d-6
+         i_max = 100
+         i_min = 0
+         num_cg_iters = 0
+
+         if (self%verbosity > 0) write(*,*) '  Running BiCG'
+      end if
+
+      call wall_time(t9)
+      do l=1, nout
```

Comment:
This cg loop feels like it could be in a function, but maybe I'm wrong

Reply:
I was split between putting this in comm_tod_mapmaking and just putting it as a function in this file

Reply:
I think function in the file is fine, just getting it out of the main loop makes the main logic a lot clearer I think